api.py

```diff
@@ -21,12 +20,121 @@
 
 def random_string(length=20):
     alphabet = string.ascii_letters + string.digits + "_/-.;:#+*?()$[]!"
-    return "".join((random.choice(alphabet) for i in range(length)))
+    return "".join((random.choice(alphabet) for _ in range(length)))
+
+
+def fetch(url: str):
+    request = tornado.httpclient.HTTPRequest(url, headers={
+        "User-Agent": ""
+    })
+    http_client = tornado.httpclient.AsyncHTTPClient()
+    return http_client.fetch(request, raise_error=True)
+
+
+@gen.coroutine
+def github_parser(repo_name: str):
+    url = "https://github.com/TeamNewPipe/{}/releases/".format(repo_name)
+    html_string = (yield fetch(url)).body
+    document = html.fromstring(html_string)
+
+    @gen.coroutine
+    def version_get() -> str:
+        tags = document.cssselect(".release .float-left ul li a.css-truncate > span.css-truncate-target")
+        return tags[0].text
+
+    gradle_template = "https://raw.githubusercontent.com/TeamNewPipe/{}/{}/app/build.gradle"
+
+    @gen.coroutine
+    def version_code_get() -> int:
+        tags = document.cssselect(".release .float-left ul li a code")
+        repo_hash = tags[0].text
+        gradle_file_data = (yield fetch(gradle_template.format(repo_name, repo_hash))).body
+        if isinstance(gradle_file_data, bytes):
+            gradle_file_data = gradle_file_data.decode()
+        version_codes = re.findall("versionCode(.*)", gradle_file_data)
+        return int(version_codes[0].split(" ")[-1])
+
+    @gen.coroutine
+    def apk_get() -> str:
+        tags = document.cssselect('.release-main-section li.d-block a[href$=".apk"]')
+        return "https://github.com" + tags[0].get("href")
+
+    return {
+        "stable": (yield gen.multi({
+            "version": version_get(),
+            "version_code": version_code_get(),
+            "apk": apk_get(),
+        }))
+    }
+
+
+@gen.coroutine
+def fdroid_parser(package_name: str):
+    template = "https://gitlab.com/fdroid/fdroiddata/raw/master/metadata/{}.yml"
+    url = template.format(package_name)
+    version_data = (yield fetch(url)).body
+    if isinstance(version_data, bytes):
```

Comment:
You shouldn't be too short on empty lines. A few more would increase readability.